db/db_impl_compaction_flush.cc

```diff
@@ -1367,40 +1415,53 @@ Status DBImpl::BackgroundFlush(bool* made_progress, JobContext* job_context,
     return status;
   }
 
-  ColumnFamilyData* cfd = nullptr;
+  autovector<ColumnFamilyData*> cfds;
+  autovector<MutableCFOptions> mutable_cf_options_list;
+  autovector<uint64_t> memtable_ids;
   while (!flush_queue_.empty()) {
     // This cfd is already referenced
-    auto first_cfd = PopFirstFromFlushQueue();
-
-    if (first_cfd->IsDropped() || !first_cfd->imm()->IsFlushPending()) {
-      // can't flush this CF, try next one
-      if (first_cfd->Unref()) {
-        delete first_cfd;
+    const auto& flush_req = PopFirstFromFlushQueue();
+
+    for (auto& iter : flush_req) {
+      auto cfd = iter.first;
+      if (cfd->IsDropped() || !cfd->imm()->IsFlushPending()) {
+        // can't flush this CF, try next one
+        if (cfd->Unref()) {
+          delete cfd;
+        }
+        continue;
       }
-      continue;
+      cfds.push_back(cfd);
+      mutable_cf_options_list.emplace_back(*cfd->GetLatestMutableCFOptions());
+      memtable_ids.push_back(iter.second);
+      job_context->superversion_contexts.emplace_back(
+          SuperVersionContext(true));
+    }
+    if (!cfds.empty()) {
+      break;
     }
-
-    // found a flush!
-    cfd = first_cfd;
-    break;
   }
 
-  if (cfd != nullptr) {
-    const MutableCFOptions mutable_cf_options =
-        *cfd->GetLatestMutableCFOptions();
+  if (!cfds.empty()) {
     auto bg_job_limits = GetBGJobLimits();
-    ROCKS_LOG_BUFFER(
-        log_buffer,
-        "Calling FlushMemTableToOutputFile with column "
-        "family [%s], flush slots available %d, compaction slots available %d, "
-        "flush slots scheduled %d, compaction slots scheduled %d",
-        cfd->GetName().c_str(), bg_job_limits.max_flushes,
-        bg_job_limits.max_compactions, bg_flush_scheduled_,
-        bg_compaction_scheduled_);
-    status = FlushMemTableToOutputFile(cfd, mutable_cf_options, made_progress,
-                                       job_context, log_buffer);
-    if (cfd->Unref()) {
-      delete cfd;
+    for (const auto& cfd : cfds) {
+      ROCKS_LOG_BUFFER(
+          log_buffer,
+          "Calling FlushMemTableToOutputFile with column "
+          "family [%s], flush slots available %d, compaction slots available "
+          "%d, "
+          "flush slots scheduled %d, compaction slots scheduled %d",
+          cfd->GetName().c_str(), bg_job_limits.max_flushes,
+          bg_job_limits.max_compactions, bg_flush_scheduled_,
+          bg_compaction_scheduled_);
+    }
+    status =
+        FlushMemTablesToOutputFiles(cfds, mutable_cf_options_list, memtable_ids,
+                                   made_progress, job_context, log_buffer);
```

Comment:
This interface is very confusing to me. Rather than doing it, can you define a struct containing ColumnFamilyData*, MutableCFOptions and memtable ID, and pass in a vector of that?